Bas beh sched engine

bjoernschroeder opened the pull request.

bjoernschroeder committed:
notify workers when job has been cancelled

bjoernschroeder committed:
add terminate method to scheduler

bjoernschroeder committed:
dont execute terminate twice

bjoernschroeder committed:
terminate scheduler when engine is terminated

bjoernschroeder committed:
add basicPlan member

bjoernschroeder committed:
add getter for basicPlan

bjoernschroeder committed:
Create basicPlan, store in plan

bjoernschroeder committed:
get basicPlan from plan

bjoernschroeder committed:
remove PlanPool

bjoernschroeder committed:
use default destructor

bjoernschroeder committed:
formatting

bjoernschroeder committed:
add const to operators

bjoernschroeder committed:
add id to jobs

bjoernschroeder committed:
use ids to compare jobs

bjoernschroeder committed:
formatting

StephanOpfer committed:
Merge branch 'rr-devel' into bas_alica_scheduling

bjoernschroeder committed:
add id to constructor args, remove unused imports

bjoernschroeder committed:
generate jobID in scheduler

bjoernschroeder committed:
return nullptr when jobQueue is empty

bjoernschroeder committed:
add popNext method

bjoernschroeder committed:
remove terminated worker threads from vector

bjoernschroeder committed:
store worker threads as objects instead of ptrs

bjoernschroeder committed:
initialize _jobID with 0

bjoernschroeder committed:
rename method to editScheduler

bjoernschroeder committed:
use std::move for scheduling, use raw ptrs for BasicPlans

bjoernschroeder committed:
use std::move to add jobs to queue

bjoernschroeder committed:
do not schedule prerequisites in schedule method

bjoernschroeder committed:
remove redundant check before scheduling

bjoernschroeder committed:
remove isScheduled

bjoernschroeder committed:
take rvalue as an argument of insert, use std::move for scheduling and inserting

bjoernschroeder committed:
use const for exception, use ALICA_ERROR_MSG for print

bjoernschroeder committed:
do not sort the entire queue when a job is inserted

bjoernschroeder committed:
remove inExecution flag from jobs

bjoernschroeder committed:
insert repeated tasks into separate queue

bjoernschroeder committed:
Merge branch 'bas_alica_scheduling' of https://github.com/rapyuta-robotics/alica into bas_alica_scheduling

bjoernschroeder committed:
store unique_ptr of BasicPlans in Plan

bjoernschroeder committed:
remove name from BasicPlan

bjoernschroeder committed:
add init method to scheduler

bjoernschroeder committed:
get ThreadPoolSize from Alica.yaml

bjoernschroeder committed:
dont overwrite terminateJobs of children on switching states, create weak_ptr before moving shared_ptr

bjoernschroeder committed:
remove mutex, insert without sorting queue

bjoernschroeder committed:
use mutex when accessing jobQueue

bjoernschroeder committed:
regenerate, add plans for skipState scheduling test

bjoernschroeder committed:
add test for scheduling with skipping states

bjoernschroeder committed:
fix enum34 issue

bjoernschroeder committed:
update plans

bjoernschroeder committed:
add getAvailableJob and allPrerequisitesFinished

bjoernschroeder committed:
use getAvailableJob

bjoernschroeder committed:
remove isEmpty method

bjoernschroeder committed:
check if scheduler has been terminated first

bjoernschroeder committed:
order jobs by scheduled time only

bjoernschroeder committed:
add isPrerequisisteFree

bjoernschroeder committed:
use uint in for loop

bjoernschroeder committed:
add prints

bjoernschroeder committed:
add prints

bjoernschroeder committed:
use isPrerequisiteFree

bjoernschroeder committed:
remove allPrerequisitesFinished method

bjoernschroeder committed:
remove prints

bjoernschroeder committed:
pass config to scheduler, get threadPoolSize from config, remove init method of scheduler

bjoernschroeder committed:
add getPrerequisites

bjoernschroeder committed:
remove isPrerequisite

bjoernschroeder committed:
remove popNext

bjoernschroeder committed:
add jobOrderingTest

bjoernschroeder committed:
add break to for loop

bjoernschroeder committed:
initialize prerequisiteIndex with 0

bjoernschroeder committed:
add jobIsPrerequisiteFree test

bjoernschroeder committed:
add test for scheduler getNextJobID

bjoernschroeder committed:
add test for scheduler getPrerequisites

bjoernschroeder committed:
add test for scheduling

bjoernschroeder committed:
use for loop for iterating through queue

bjoernschroeder committed:
add test for jobQueue

bjoernschroeder committed:
remove blank lines

bjoernschroeder committed:
add comments

bjoernschroeder committed:
remove old code

bjoernschroeder committed:
compare size of prerequisites vector with unsigned int value

StephanOpfer committed:
Merge branch 'rr-devel' into bas_alica_scheduling

bjoernschroeder committed:
add getInterval

bjoernschroeder committed:
add runJob

bjoernschroeder committed:
add void* parameter to callbacks, add inProgress boolean

bjoernschroeder committed:
schedule and cancel runJobs

bjoernschroeder committed:
dont remove repeated jobs from queue

bjoernschroeder committed:
store lowest scheduled time

bjoernschroeder committed:
wake up workers when job is ready for execution, set inProgress and scheduledTime for jobs

bjoernschroeder committed:
add monitor thread for starving jobs

bjoernschroeder committed:
dont delete cancelled jobs which are in progress

bjoernschroeder committed:
add isEmpty

bjoernschroeder committed:
add getLowestScheduledTime

bjoernschroeder committed:
dont execute run method in behaviour thread

bjoernschroeder committed:
add test for repeatedJobs, create callbacks with void* parameter

bjoernschroeder committed:
take rvalue of jobs

bjoernschroeder committed:
use std::move when scheduling jobs

bjoernschroeder committed:
add comment to getPrerequisites

bjoernschroeder committed:
use return instead of continue when running is set to false

bjoernschroeder committed:
store scheduler as ptr

bjoernschroeder committed:
remove engine from scheduler, store AlicaClock in scheduler

bjoernschroeder committed:
prevent execution in BehaviourThread, replace doRun, change visibility of behaviour init, terminate and run method to public

bjoernschroeder committed:
schedule behaviour init, terminate, run and wait

bjoernschroeder committed:
remove void* as cb argument

bjoernschroeder committed:
prevent start of behaviour

bjoernschroeder committed:
dont return repeated task in execution

bjoernschroeder committed:
execute cb without nullptr argument

bjoernschroeder committed:
create cb without void* argument, remove repeatedJobExecution test

bjoernschroeder committed:
add timeouts to simple plan test

bjoernschroeder committed:
add behaviour trigger support, refactoring

bjoernschroeder committed:
regenerate files

bjoernschroeder committed:
set deactivated siblings terminate methods as prerequisites for behaviours

bjoernschroeder committed:
implement rulebook tests

bjoernschroeder committed:
add plans for rulebook tests

bjoernschroeder committed:
update plan files

bjoernschroeder committed:
update behaviour trigger test

bjoernschroeder committed:
add deactivated children as prerequisite, remove debug prints

bjoernschroeder committed:
pass deactivated children to rp in rules

bjoernschroeder committed:
add flags for preconditions

bjoernschroeder committed:
add counter to empty behaviour

bjoernschroeder committed:
implement summand

bjoernschroeder committed:
regenerate plans

bjoernschroeder committed:
implement remaining tests for rules

bjoernschroeder committed:
add test for behaviour frequency

bjoernschroeder committed:
fix lowest scheduled time calculation

veerajsk committed:
Add timer manager

veerajsk committed:
Add callbackq & threadpool template params

bjoernschroeder committed:
create and run timers for run jobs

bjoernschroeder committed:
update plans

bjoernschroeder committed:
Merge remote-tracking branch 'origin/vsk_ros_timer_beh_sched' into bas_behaviour_scheduling

bjoernschroeder committed:
start threadpool on creation, return ptr to AlicaTimer, fix minor errors

bjoernschroeder committed:
simplify JobQueue

bjoernschroeder committed:
remove lowestScheduledTime and mtx

bjoernschroeder committed:
return timer as unique_ptr, wrap mtx and condition_variable in unique_ptr

bjoernschroeder committed:
schedule runJob in doInit, stop runJob in onTermination

bjoernschroeder committed:
remove scheduling of runJobs

bjoernschroeder committed:
create, stop and manage runJob timers, execute init/terminate single threaded

bjoernschroeder committed:
expect at least 30 executions of runMethod

bjoernschroeder committed:
remove prerequisites, remove cancelled and inProgress flag

bjoernschroeder committed:
return AlicaTimer objects by value

bjoernschroeder committed:
stop timers when scheduler is terminated

bjoernschroeder committed:
add sleep to tests

bjoernschroeder committed:
Merge branch 'bas_alica_scheduling' into bas_behaviour_scheduling

# Conflicts:
#	alica_engine/include/engine/scheduler/Scheduler.h
#	alica_engine/src/engine/scheduler/Scheduler.cpp

bjoernschroeder committed:
take ownership of passed job, pass AlicaClock instead of engine, remove unused members

bjoernschroeder committed:
remove unused methods, formatting

bjoernschroeder committed:
remove unused tests

bjoernschroeder committed:
init rostimer in constructor and move constructor, implement move constructor

bjoernschroeder committed:
move timer into map

bjoernschroeder committed:
remove job ordering and scheduled time

bjoernschroeder committed:
remove unused members of job

bjoernschroeder committed:
remove alicaClock, take repeatInterval as input, change getNextJobId to private

bjoernschroeder committed:
dont pass alicaClock to scheduler

bjoernschroeder committed:
update job creation

bjoernschroeder committed:
schedule behaviour init on behaviour start

bjoernschroeder committed:
work on stopping and terminating behaviour

veerajsk committed:
Refactor

bjoernschroeder committed:
store scheduler as value

bjoernschroeder committed:
create init and terminate jobs in BasicPlan

bjoernschroeder committed:
create terminate job in BasicBehaviour

bjoernschroeder committed:
remove job creation from RunningPlan

bjoernschroeder committed:
remove timerMtx

bjoernschroeder committed:
Merge remote-tracking branch 'origin/vsk_refactor_bas_scheduling' into bas_behaviour_scheduling

# Conflicts:
#	alica_engine/include/engine/scheduler/Scheduler.h

bjoernschroeder committed:
fixes for refactoring

bjoernschroeder committed:
fix cv

veerajsk committed:
Fix code

veerajsk committed:
Merge pull request #145 from rapyuta-robotics/vsk_fix_bas_scehduling

Fix code

veerajsk committed:
Move value while popping the job queue

bjoernschroeder committed:
change c++ version to 17

bjoernschroeder committed:
fix scheduler

bjoernschroeder committed:
fix behaviour trigger

bjoernschroeder committed:
add test for plan init/terminate

veerajsk committed:
Revert "Fix code"

This reverts commit 71b165874b9432c0ab550a763e5f1cf6d07b6fb6.

veerajsk committed:
Redo the non-formatting related changes

veerajsk committed:
Merge branch 'rr-devel' into bas_behaviour_scheduling

veerajsk committed:
Revert alica test changes

veerajsk committed:
Update tests

bjoernschroeder committed:
remove dynalloc utility function

abhi1287 commented on alica_engine/CMakeLists.txt:
```diff
@@ -10,16 +10,16 @@ find_package(yaml-cpp REQUIRED)
 find_package(catkin QUIET)
 if (catkin_FOUND)
-  find_package(catkin REQUIRED COMPONENTS id_manager fsystem event_handling system_util alica_common_config alica_solver_interface)
+  find_package(catkin REQUIRED COMPONENTS roscpp id_manager fsystem event_handling system_util alica_common_config alica_solver_interface)
```
engine shouldn't depend on ros. alica applications can.

veerajsk replied:
yeah, that was the next step in this PR. I have resolved it as discussed

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -160,5 +165,7 @@ class BasicBehaviour
     mutable std::mutex _runLoopMutex;
     AlicaTime _msInterval;
     AlicaTime _msDelayedStart;
+    int64_t _activeRunJobId;
```
Can remove _runThread now.

abhi1287 commented on alica_engine/include/engine/AlicaTimer.h:
```diff
@@ -0,0 +1,146 @@
+    TimerCb _userCb;
+    ros::NodeHandle _nh;
+    ros::Timer _timer;
+    std::shared_ptr<CallbackQ> _cbQ;
```
Scheduler can make sure timerFactory lifetime is more than any of the timers and thread pool. We can then pass _cbQ around with reference.

veerajsk replied:
I had it this way in the beginning & then changed to using shared_ptrs. If we are okay to make sure timerfactory has a larger life time, I guess its ok to revert to using plain references.

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -0,0 +1,24 @@
+    virtual void onTermination(){};
+    void start();
+    void stop();
+    void setEngine(AlicaEngine* engine) { _ae = engine; }
```
need to set running plan context and a getter to get it. just like in basicbehaviour.

veerajsk replied:
done

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -0,0 +1,113 @@
+
+    JobId schedule(JobCb jobCb, std::optional<AlicaTime> repeatInterval = std::nullopt)
+    {
+        _jobQueue.emplace(_nextJobId++, std::move(jobCb), std::move(repeatInterval));
```
_nextJobId should be incremented atomically

veerajskhokale replied:
done

abhi1287 commented on alica_engine/include/engine/RunningPlan.h:
```diff
@@ -222,6 +225,7 @@ class RunningPlan : public std::enable_shared_from_this<RunningPlan>
     RunningPlan* _parent;
 
     BasicBehaviour* _basicBehaviour;
+    BasicPlan* _basicPlan;
```
can use std::variant

abhi1287 replied:
bump

veerajsk committed:
Address review comments

bjoernschroeder committed:
remove runThread

bjoernschroeder committed:
add context and config

bjoernschroeder committed:
make nextJobId atomic

bjoernschroeder committed:
update job creation, add setConfiguration

bjoernschroeder committed:
set runningPlan and config

veerajsk committed:
Fixes

- set engine during BasicPlan creation
- fix compile error in setting running plan context of BasicPlan
- removed some minor redundant code

veerajsk committed:
Use the same BasicBehaviour object for every configuration

bjoernschroeder committed:
fix context

bjoernschroeder committed:
Merge branch 'vsk_beh_sched_fixes' of github.com:rapyuta-robotics/alica into vsk_beh_sched_fixes

veerajsk committed:
Revert "Use the same BasicBehaviour object for every configuration"

This reverts commit 16b50027f8b5ecfef41a83ecd91898cc7171d1d0.

veerajsk committed:
Fix build error

veerajsk committed:
Add run method to basic plan

veerajsk committed:
Fix compile error

bjoernschroeder committed:
set timerFactory

bjoernschroeder committed:
fix typo, add alica_ros_timer as dependency

veerajsk committed:
Construct scheduler in init since timer factory won't be available

veerajsk committed:
Merge pull request #151 from rapyuta-robotics/vsk_beh_sched_fixes

Vsk beh sched fixes

bjoernschroeder committed:
check if scheduler is not nullptr before terminating

bjoernschroeder committed:
get ThreadPoolSize from config

abhi1287 commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -289,6 +293,30 @@ class AlicaContext
+
+    /**
+     * Set timer factory to be used by this alica framework instance.
+     * Example usage: setCommunicator<alicaRosTimer::TimerFactoryRos>();
```
update comment

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -75,6 +75,15 @@ class BasicBehaviour
+    AlicaTime getInterval() { return _msInterval; }
+
+    bool isEventDriven() const { return _behaviour->isEventDriven(); }
+
```
looks like all these methods could be private.

abhi1287 replied:
A lot of functions here can still be moved to private or protected region.

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -0,0 +1,52 @@
+    virtual void run(void* msg){};
+    virtual void onTermination(){};
+
+    void doInit();
```
similar to basicBehaviour, these methods could be private

abhi1287 commented on alica_engine/include/engine/IAlicaTimer.h:
```diff
@@ -0,0 +1,28 @@
+
+class IAlicaTimerFactory
+{
+public:
```
Looks like we don't need to pass threadpool size from engine anymore?

veerajsk replied:
yes, the thread pool size can be passed by the application when it sets the timer factory via the context

abhi1287 commented on alica_engine/include/engine/scheduler/JobQueue.h:
```diff
@@ -0,0 +1,54 @@
+private:
+    std::mutex _mutex;
+    std::deque<T> _queue;
+};
```
AlicaDummyCommunication also has a InProcQueue. In a different pr we can modify it to use this one.

abhi1287 commented on alica_engine/include/engine/IAlicaTimer.h:
```diff
@@ -0,0 +1,28 @@
+{
+public:
+    virtual std::unique_ptr<IAlicaTimer> createTimer(IAlicaTimer::TimerCb timerCb, AlicaTime period) = 0;
+    virtual ~IAlicaTimerFactory() = default;
```
Need proper comments on the interface methods.

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -0,0 +1,113 @@
+    {
+        _jobQueue.emplace(_nextJobId++, std::move(jobCb), std::move(repeatInterval));
+        _cv.notify_one();
+        return _nextJobId;
```
Take _nextJobId++ into a local variable and use that. _nextJobId can change by the time this line is hit.

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -0,0 +1,113 @@
+        _repeatableJobs.clear();
+    }
+
+    void stopJob(JobId jobId) { _repeatableJobs.erase(jobId); }
```
This function should be renamed to cancelRepeatedJob(). I see a race condition here, what if the job is still in _jobQueue ? This cancel will be a no-op and scheduler will start the job later. 
One option is to have a cancelled flag for Job and not process it when the time comes.

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -0,0 +1,113 @@
+            }
+            auto&& [jobId, jobCb, repeatInterval] = std::move(*job);
+            if (repeatInterval) {
+                _repeatableJobs[jobId] = _timerFactory.createTimer(std::move(jobCb), std::move(*repeatInterval));
```
_repeatableJobs needs to be mutex protected

abhi1287 replied:
Maybe you can remove mutex from FIFO and rely on just the scheduler mutex.

abhi1287 commented on alica_engine/package.xml:
```diff
@@ -16,6 +16,7 @@
   <depend>event_handling</depend>
   <depend>system_util</depend>
   <depend>id_manager</depend>
+  <depend>roscpp</depend>
```
remove

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,14 +84,12 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
-    }
-    if (_behaviour->isEventDriven()) {
-        _runCV.notify_all();
+    if (!isStopCalled()) {
```
Don't need to check if (!isStopCalled()) . If a running plan instance called stop that means it must have called a start before.

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,14 +84,12 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
     }
+
+    setSignalState(SignalState::STOP);
+    setStopCalled(true);
```
Too many trackers for state. As @veerajsk  you said before, we definitely need to clean this up in a different pr.

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -200,61 +173,27 @@ void BasicBehaviour::doStop()
-            _behaviourTrigger->setNotifyCalled(&_runCV, false);
-        }
-    }
+    run(msg);
```
catch exceptions from this, like it was done before.

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -0,0 +1,54 @@
+void BasicPlan::doInit()
+{
+    _planStarted = true;
+    init();
```
catch exceptions from init, run and terminate.

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -0,0 +1,54 @@
+void BasicPlan::doTerminate()
+{
+    _ae->editScheduler().stopJob(_activeRunJobId);
+    onTermination();
```
@bjoernschroeder  since we are starting fresh, lets name methods onInit(), onTerminate()

abhi1287 commented on alica_engine/src/engine/RunningPlan.cpp:
```diff
@@ -557,7 +560,14 @@ void RunningPlan::activate()
     _status.active = PlanActivity::Active;
     if (isBehaviour()) {
         _ae->editBehaviourPool().startBehaviour(*this);
+        _basicPlan = nullptr;
```
this should be in constructor

bjoernschroeder committed:
update comment

bjoernschroeder committed:
add doc comments for interface methods

bjoernschroeder committed:
use local variable for nextJobId

bjoernschroeder committed:
remove ros dependency

bjoernschroeder committed:
init basicPlan with nullptr

veerajsk committed:
Rewrite basic behaviour

bjoernschroeder committed:
implement cancelJob

bjoernschroeder committed:
remove setTrigger

bjoernschroeder committed:
catch exceptions

bjoernschroeder committed:
make onTerminate, onInit and run private, rename methods

bjoernschroeder committed:
rename onInit and onTerminate behaviour methods

bjoernschroeder committed:
regenerate files

veerajsk committed:
Merge pull request #155 from rapyuta-robotics/bas_review_fixes

Bas review fixes

veerajsk committed:
Merge pull request #152 from rapyuta-robotics/bas_beh_sched_fixes

check if scheduler is not nullptr before terminating

veerajsk committed:
__testing

veerajsk committed:
__testing again

veerajsk committed:
add plans

veerajsk committed:
gen code

veerajsk committed:
Revert "__testing"

This reverts commit 960df92f897f82237d1ab70398cbdc103a790b31.

veerajsk committed:
Revert "__testing again"

This reverts commit 8aebe97bb2a6a488fdd3a8832364a903b2c03879.

veerajsk committed:
Init/term methods for plans/beh

veerajsk committed:
Add unit test for init/term execution order

veerajsk committed:
Add unit test for run order verification

veerajsk committed:
Add unit test for success/failure semantics

veerajsk committed:
Check if executing in context before setting beh success/failure

veerajsk committed:
Add alica_ros_timer to catkin depends

bjoernschroeder committed:
set frequency for plans

veerajsk committed:
Remove start, stop methods for alica timers

Start & stop will be done on construction & destruction instead

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,168 +56,108 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
+    // Do not schedule repeatable run job when behaviour is event driven.
+    if (!isEventDriven()) {
+        // TODO: account for delayed start
+        _activeRunJobId = _engine->editScheduler().schedule(std::bind(&BasicBehaviour::runJob, this, nullptr), getInterval());
```
if frequency is 0 don't schedule run job -> especially important for BasicPlan

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,168 +56,108 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
-    while (doWait()) {
-        doInit(timed);
-        doRun(timed);
+    _engine->editScheduler().cancelJob(_activeRunJobId);
```
check if run job is scheduled in the first place before calling cancel

abhi1287 replied:
not required anymore

veerajsk replied:
it is still required for behaviour's that have frequency 0

bjoernschroeder committed:
set frequency for basicPlans

bjoernschroeder committed:
dont schedule runjob of plans with frequency 0

bjoernschroeder committed:
set interval of to -1 when freq is 0

bjoernschroeder committed:
dont cancle run job if no run job has been scheduled

veerajsk commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -0,0 +1,53 @@
+#pragma once
```
this file has to be rewritten in the same way as BasicBehaviour. In its current state getPlanContext() won't work as expected & may lead to dereferencing a deleted RunningPlan

abhi1287 replied:
Do you want to do it in different pr? 
Copying start / stop / init / terminate api's look trivial with related isExecutingInContext() logic

veerajsk replied:
I'll do it in the same one, infact @bjoernschroeder already has made the changes, will merge it after review

veerajsk committed:
Add alica-supplementary to the ros dependencies

bjoernschroeder committed:
add execState and signalState to BasicPlan

bjoernschroeder committed:
pass RP in start, fix returns

veerajsk commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -37,34 +37,34 @@ class BasicBehaviour
 
-    ThreadSafePlanInterface getPlanContext() const { return ThreadSafePlanInterface(isBehaviourStarted() ? _context : nullptr); }
-    void setRunningPlan(RunningPlan* rp) { _context = rp; }
+    ThreadSafePlanInterface getPlanContext() const { return ThreadSafePlanInterface(isExecutingInContext() ? _context.load() : nullptr); }
```
It does not seem possible to guarantee that the pointer returned by getPlanContext() is safe to access in the behaviour unless a shared_ptr is used to ensure that the RunningPlan is not destroyed by the alica main engine thread. Even using a mutex here would not help

veerajsk commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,168 +56,108 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
+        return true;
     }
+    ++_signalState;
+    // This has to be done after incrementing _signalState for correct behaviour of getPlanContext()
```
Better to remove this comment since it is not entirely accurate. See getPlanContext()'s comment in the header file

veerajsk committed:
Move test code outside autogenerated code

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -92,44 +97,24 @@ class BasicBehaviour
+    void initJob();
+    void terminateJob();
+
+    static constexpr bool isActive(BehState st) { return !(st & 1); }
```
As discussed over call, please add some comments about how we are tracking things here.

veerajsk replied:
yes, will do

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -92,44 +97,24 @@ class BasicBehaviour
 private:
     friend alica::test::TestContext;
-    enum BehaviourResult
+    using BehState = uint64_t;
```
BehState sounds misleading. Can be called simple counter.

veerajsk replied:
okay

abhi1287 commented on alica_engine/include/engine/IAlicaTimer.h:
```diff
@@ -0,0 +1,38 @@
+     * The implementation of stop should stop the repeated execution of the TimerCb.
+     * Stopping the timer is expected to be synchronized. After leaving the stop method,
+     * the timer should no longer execute any TimerCb. If TimerCb is already invoked stop
+     * waits until TimerCb returns, therefore destruction is blocking.
```
Can simplify to:
The destructor should stop the repeated execution of the TimerCb and needs to block if any TimerCb is already in progress.

abhi1287 commented on alica_engine/include/engine/expressionhandler/ExpressionHandler.h:
```diff
@@ -17,7 +18,7 @@ class ExpressionHandler
     ExpressionHandler();
     virtual ~ExpressionHandler();
-    void attachAll(PlanRepository& pr, AlicaCreators& creatorCtx);
+    void attachAll(AlicaEngine* ae, PlanRepository& pr, AlicaCreators& creatorCtx);
```
Please put a TODO, to be reverted once we have PlanPool from Bjoern.

abhi1287 commented on alica_engine/include/engine/model/Plan.h:
```diff
@@ -77,6 +80,7 @@ class Plan : public AbstractPlan
     FailureStateGrp _failureStates;
     SynchronisationGrp _synchronisations;
     TransitionGrp _transitions;
+    std::unique_ptr<BasicPlan> _basicPlan;
```
to be moved to PlanPool

abhi1287 commented on alica_engine/include/engine/scheduler/Scheduler.h:
```diff
@@ -0,0 +1,123 @@
+    void cancelJob(JobId jobId)
+    {
+        // Should be called by the scheduler thread
+        if (_repeatableJobs.erase(jobId)) {
```
if (!_repeatableJob.erase(jobId))

veerajsk replied:
looks like unit tests can be further improved, because they didn't catch this bug

abhi1287 commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -111,6 +115,9 @@ void AlicaEngine::start()
 {
     _maySendMessages = false;
+    if (_scheduler) {
+        _scheduler->terminate();
```
guess for consistency it can have an init method that will start the scheduler thread.

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,168 +56,108 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
-    _behaviourTrigger->registerCV(&_runCV);
+    }
+    auto prev = _behResult.exchange(BehResult::FAILURE);
+    if (prev != BehResult::SUCCESS) {
```
FAILURE

abhi1287 commented on alica_tests/autogenerated/src/Behaviour/NotToTrigger.cpp:
```diff
@@ -18,7 +18,6 @@ NotToTrigger::NotToTrigger()
     // Add additional options here
     this->callCounter = 0;
     this->initCounter = 0;
-    setTrigger(alicaTests::TestWorldModel::getOne()->trigger1.get());
```
test_alica_behaviourtrigger will now fail. Since it uses alicaTests::TestWorldModel::getOne()->trigger1->run(true); And you removed the dependency of ITrigger. Now it is cleaner but you will have to modify the trigger testcases to call doTrigger() instead of ITrigger->run()

veerajsk replied:
yes, that should be done already in https://github.com/rapyuta-robotics/alica/pull/146

veerajsk committed:
Address review comments

- Add comments explaining basic behaviour logic
- Fix comment for timer interface
- Add init() method for scheduler for consistency
- Bug fix: check for failure in setFailure(), not success
- Add TODO comments for code that should be replaced after plan pools
are implemented
- Bug fix: remove from job queue if timer is not scheduled for the job
- Tighter access control for functions in basic behaviour
- More intuitive name for counters used to track behaviour execution

veerajsk committed:
Set the running variable before starting the scheduler thread

veerajsk committed:
Make sure to take the right transistion

bjoernschroeder committed:
fix PartialAssignment compare

veerajsk committed:
Merge pull request #159 from rapyuta-robotics/bas_partial_assignment_compare_fix

fix PartialAssignment compare

veerajsk committed:
Address review comments

veerajsk committed:
Merge pull request #157 from rapyuta-robotics/bas_plan_freq

Plan Frequency

veerajsk committed:
Remove RolesetTA for importing to new designer

Commit:
update plans using the new web based plan designer

veerajsk committed:
Gen code use new designer runtime

veerajsk committed:
Minor refactor

- tighter access control for functions
- more intuitive variable names
- move small functions to header if possible for better inlining

veerajsk committed:
Merge pull request #158 from rapyuta-robotics/bas_plan_exec_signal_state

BasicPlan exec & signal state

bjoernschroeder committed:
skip roles which are not in the current roleset

bjoernschroeder committed:
clear roleTaskReferences when modelManager is set

bjoernschroeder committed:
update roleids

veerajsk committed:
Merge pull request #146 from rapyuta-robotics/bas_beh_sched_tests

Update & add tests for plan/beh scheduling

veerajsk committed:
__temp

veerajsk committed:
regenerate code

veerajsk committed:
Clear static references

veerajsk committed:
Fix role id

veerajsk committed:
Beh Freq 0 -> 1

veerajsk committed:
Add unit test to check run method

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -37,53 +37,56 @@ class BasicBehaviour
+    bool start(RunningPlan* rp);
+    void terminate() { stop(); };
 
     void setDelayedStart(int32_t msDelayedStart) { _msDelayedStart = AlicaTime::milliseconds(msDelayedStart); }
```
This is a no-op now. Can remove.

abhi1287 commented on alica_engine/include/engine/BasicPlan.h:
```diff
@@ -0,0 +1,66 @@
+    void start(RunningPlan* rp);
+    void stop();
+
+    void setEngine(AlicaEngine* engine) { _ae = engine; }
```
could be given in constructor

abhi1287 commented on alica_engine/include/engine/model/Plan.h:
```diff
@@ -99,6 +106,10 @@ class Plan : public AbstractPlan
      */
     PreCondition* _preCondition;
+    /**
+     * The frequency with which this Plan is called.
```
run method is called

abhi1287 commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -83,12 +85,15 @@ void AlicaEngine::reload(const YAML::Node& config)
  */
 bool AlicaEngine::init(AlicaCreators& creatorCtx)
 {
+    _scheduler = std::make_unique<scheduler::JobScheduler>(_ctx.getTimerFactory());
```
_scheduler can be just a normal member.

abhi1287 commented on alica_engine/src/engine/AlicaEngine.cpp:
```diff
@@ -111,6 +116,9 @@ void AlicaEngine::start()
 void AlicaEngine::terminate()
 {
     _maySendMessages = false;
+    if (_scheduler) {
```
won't have to check this if its a member

abhi1287 commented on alica_engine/src/engine/BasicBehaviour.cpp:
```diff
@@ -92,168 +56,110 @@ essentials::IdentifierConstPtr BasicBehaviour::getOwnId() const
-    if (!_behaviour->isEventDriven()) {
-        _runCV.notify_all();
+    if (isActive(_signalState.load())) {
+        return true;
```
put an assert in start in "else " of this condition. Stop can be called repeatedly during termination.

abhi1287 commented on alica_engine/src/engine/BasicPlan.cpp:
```diff
@@ -0,0 +1,86 @@
+{
+    if (_activeRunJobId != -1) {
+        _ae->editScheduler().cancelJob(_activeRunJobId);
+        _activeRunJobId = -1;
```
have -1 defined as const InvalidJobId

abhi1287 reviewed (approved):
nothing blocking, can be taken in different pr

veerajsk commented:
> nothing blocking, can be taken in different pr

sure, @bjoernschroeder can address the review comments in a subsequent PR

veerajsk committed:
Merge branch 'rr-devel' into bas_beh_sched_engine
